### PR TITLE
🧹 Clean up jquery rails

### PR DESCRIPTION
**Pull request by josemigallas**
### TL;DR
**UPDATE**: It was not so simple!

### Description
Forms and other asynchronous stuff is managed by `rails-ujs` (Unobtrusive scripting adapter for jQuery).

It is embedded in the jQuery imported as gem `jquery-rails` as following:
```ruby
//= require jquery
//= require jquery_ujs
```

In recent rails (5+), rails-ujs is included in the gem rails and it's imported like this:
```ruby
//= require jquery
//= require rails-ujs
```

On a side note `jquery-rails` includes jquery 1, 2 and 3:
```ruby
//= require jquery
//= require jquery2
//= require jquery3
```

I think the best way to unify jquery is to remove all versions from assets/vendor and just use the one provided by this gem.

Webpack packages will still use jQuery 3.x from node_modules and there's even a node module for rails-ujs... so in time we'll see wether we stick to Gem of NPM.

### CODE REVIEW

File: lib/developer_portal/lib/liquid/filters/rails_helpers.rb [hunk 1]
```diff
@@ -238,7 +238,7 @@ def self.replace_googleapis(url)
         return url unless Rails.env.test?
 
         if url =~ %r{//ajax.googleapis.com/ajax/libs/jquery/(.+?)/jquery.min.js}
-          "vendor/jquery-#{$1}.min.js"
+          "essential_assets/jquery-#{$1}.min.js"
```
Comment (mayorova) [inline lib/developer_portal/lib/liquid/filters/rails_helpers.rb:241]: OMG, why did we have googleapis in the first place??!

Looks like there is still some leftover: https://github.com/3scale/porta/blob/36eeb457a682b3a1e96d99cdf91738551bf4482c/app/views/stats/days/show.html.erb#L1

I think in the next PR we'd need to remove both this helper, and that external JS.

**Review by mayorova — DISMISSED** [review]

**Review by jlledom — DISMISSED** [review]
Works apparently fine for me 👍

**Review by jlledom — APPROVED** [review]
Too much to review. If tests pass, I'm fine 👍

**Comment by mayorova** [comment]
I'm a bit confused by the description...

> In recent rails (5+), rails-ujs is included in the gem rails and it's imported like this:

So, is it in `rails` gem? :thinking:  Then why do we need the `jquery-rails` gem?...

> Moreover `rails-jquery` includes jquery 1, 2 and 3:

What is `rails-jquery`? another gem? Or is it a typo and you meant `jquery-rails` ?..

File: app/controllers/api/metric_visibilities_controller.rb [hunk 2]
```diff
@@ -10,33 +10,56 @@ class Api::MetricVisibilitiesController < FrontendController
   def toggle_visible
     @metric.toggle_visible_for_plan(@plan)
 
+    errors = @metric.errors.presence
+    if errors
+      flash[:error] = errors.full_messages.to_sentence
+    else
+      metric = @metric.method_metric? ? 'Method' : 'Metric'
```
Comment (mayorova) [inline app/controllers/api/metric_visibilities_controller.rb:17]: The amount of this snippet repeating throught the code (not only in this file) is crazy. I don't get why it's not a method on the `Metric` model :shrug: 

But leave it for another PR...

File: app/controllers/api/metric_visibilities_controller.rb [hunk 3]
```diff
@@ -10,33 +10,56 @@ class Api::MetricVisibilitiesController < FrontendController
   def toggle_visible
     @metric.toggle_visible_for_plan(@plan)
 
+    errors = @metric.errors.presence
```
Comment (mayorova) [inline app/controllers/api/metric_visibilities_controller.rb:13]: It's a bit hard to understand these changes... 

I guess you're replacing what was previously implemented in `js.erb` with plain controller logic?

Comment (josemigallas) [reply app/controllers/api/metric_visibilities_controller.rb:13]: Yes it's basically removing jQuery from those js.erb and moving the logic to server-side instead of doing stuff in the client.

File: app/views/api/featurings/change.js.erb [hunk 4]
```diff
@@ -1,4 +1,8 @@
+<%# If EventTarget is replaceWith'ed immediately, ajax:success and ajax:complete are not called and the spinner will stay %>
```
Comment (mayorova) [inline app/views/api/featurings/change.js.erb:1]: I find this comment (and other similar ones) a bit confusing... 
What's the EventTarget? and why is there a reference to `replaceWith` if it's not used in the snippet?

Comment (josemigallas) [reply app/views/api/featurings/change.js.erb:1]: I struggled to write it as clear as possible, I worry I won't understand it myself in the future 😓.

[EventTarget](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget) is the Element that will receive the event.

Many remote elements will change themselves or be replaced by a `.js.erb` returned  by server. This happens after `ajax:send` but before `ajax:success` and `ajax:complete`. When the latter events are to be dispatched, the original element that called AJAX (the EventTarget) does not exist anymore and therefore they are lost.

This affects specifically to the ajax-in-progress spinner: it's always added but sometimes not removed.

`replaceWith` is the jQuery method, that I decided to remove after adding all these comments. I will update them.

File: app/views/api/metric_visibilities/change.js.erb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  const newRow = '<%= escape_javascript(render(partial: "api/metrics/metric", locals: { metric: @metric })) %>';
+  oldRow.insertAdjacentHTML('afterend', newRow);
+  oldRow.remove();
+});
+
+<% if @usage_limits %>
+  void function () {
+    const slot = document.getElementById('<%= dom_id(@metric) %>_slot');
+
+    if (slot !== null && slot.querySelector('#usage_limits_table')) {
```
Comment (mayorova) [inline app/views/api/metric_visibilities/change.js.erb:13]: What would happen if the slot exists, but there is not usage table inside? :upside_down_face: 

Looks like in the original implementation the slot was always re-generated?... :thinking: 

So far I think the slot is always generated correctly.

The weird thing here is that the "slot" is generated as 
`<tr><td class="metric_slot" ...></td></tr>`

So, when it's deleted, only the `td` is removed, but the `tr` stays forever. But I see that it has been like this before, so OK.

File: app/views/api/metric_visibilities/change.js.erb [hunk 6]
```diff
@@ -0,0 +1,19 @@
+<%# If EventTarget is replaceWith'ed immediately, ajax:success and ajax:complete are not called and the spinner will stay %>
+setTimeout(() => {
+  const oldRow = document.getElementById('metric_<%= @metric.id %>');
+  const newRow = '<%= escape_javascript(render(partial: "api/metrics/metric", locals: { metric: @metric })) %>';
```
Comment (mayorova) [inline app/views/api/metric_visibilities/change.js.erb:4]: @josemigallas so, what's the difference between 
```
render(partial: "some/path", locals: { object: @object })
```
and
```
render("some/path", object: @object)
```
? in some cases one is used, and in other cases another one.

File: app/views/api/features/create.js.erb [hunk 7]
```diff
@@ -1,6 +1,9 @@
-var html = "<%= escape_javascript(render :partial => 'feature', :locals => {:feature => @feature}) %>";
 jQuery.colorbox.close();
-jQuery("table#features tbody").append(html);
-jQuery("table#features tr.notice").hide();
+
+document.querySelector('table#features tbody').innerHTML += '<%= escape_javascript(render partial: 'feature', locals: { feature: @feature }) %>';
+
+<% if @plan.issuer.features.size == 1 %>
```
Comment (mayorova) [inline app/views/api/features/create.js.erb:5]:
```suggestion
<% if @plan.issuer.features.any? %>
```

or `.size > 0`. I think `== 1` is too restrictive...

File: app/javascript/src/Common/ajaxEvents.ts [hunk 8]
```diff
@@ -1,26 +1,66 @@
...
 
-const ajaxSpinnerId = 'ajax-in-progress'
-
-jQuery1(document)
-  .on('ajax:before', () => {
-    $('body').append(`<div id="${ajaxSpinnerId}"><img src="/assets/ajax-loader.gif"></div>`)
+/**
+ * rails-ujs ajax events.
+ */
+$(document)
```
Comment (mayorova) [inline app/javascript/src/Common/ajaxEvents.ts:23]: I guess all the `console.log`s here were for debugging and need to be removed?

File: app/javascript/packs/provider.ts [hunk 9]
```diff
@@ -20,7 +19,7 @@ document.addEventListener('DOMContentLoaded', () => {
     .forEach(form => {
       form.addEventListener('change', () => {
         if (form.dataset.remote) {
-          void jQuery1.rails.handleRemote(jQuery1(form))
+          void window.Rails.handleRemote(jQuery1(form))
```
Comment (mayorova) [inline app/javascript/packs/provider.ts:22]: this `Rails` thing is something UJS adds?

File: app/javascript/src/Types/rails-ujs.d.ts [hunk 10]
```diff
@@ -0,0 +1,54 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+/* eslint-disable @typescript-eslint/naming-convention */
+/**
+ * SPOILER ALERT: As of rails 7 UJS is deprecated. It is recommended to use Turbo now.
```
Comment (mayorova) [inline app/javascript/src/Types/rails-ujs.d.ts:4]: :smiling_face_with_tear:

File: app/views/api/alerts/destroy.js.erb [hunk 11]
```diff
@@ -1,3 +0,0 @@
-$(function(){
```
Comment (mayorova) [inline app/views/api/alerts/destroy.js.erb:1]: I would actually keep this file (rewritten in JS if needed), and fix the Delete button by adding `remote: true`.
It used to be an AJAX action, but it was changed here: https://github.com/3scale/porta/pull/3636/files#diff-ab8780681158e67f54020924c99343f3c627228991ea13a9956221128a8615af
(not sure whether deliberately or not)

File: app/views/buyers/accounts/show.html.slim [hunk 12]
```diff
@@ -35,12 +35,11 @@ div class="pf-l-grid pf-m-gutter pf-m-all-6-col"
...
   div class="pf-l-grid__item"
     div class="pf-l-flex pf-m-column pf-m-flex-1"
-      - if account_plans_management_visible?
+      - if account_plans_management_visible? && plan = @account.bought_account_plan
         div class="pf-l-flex__item"
           div class="pf-c-card"
             div class="pf-c-card__body"
-              = render partial: 'account_plan', locals: { plan: @account.bought_account_plan,
-                                                          contract:  @account.bought_account_contract }
+              = render partial: 'account_plan', locals: { plan: plan,contract:  @account.bought_account_contract }
```
Comment (mayorova) [inline app/views/buyers/accounts/show.html.slim:42]:
```suggestion
              = render partial: 'account_plan', locals: { plan: plan, contract: @account.bought_account_contract }
```

or even `{ plan:, contract: @account.bought_account_contract }`

File: app/views/finance/provider/line_items/destroy.js.erb [hunk 13]
```diff
@@ -1,7 +1,15 @@
...
+    document.getElementById('<%= dom_id(@line_item) %>').remove();
+  });
 
-$('#invoice_cost').html('<%= price_tag(@invoice.cost) %>');
-$('#invoice_cost_without_vat').html('<%= price_tag(@invoice.cost(vat_included: false)) %>');
+  <% unless @invoice.vat_rate.nil? %>
+    document.getElementById('invoice_cost_without_vat').innerHTML = '<%= price_tag(@invoice.cost(vat_included: false)) %>';
+  <% end %>
+
+  document.getElementById('invoice_cost').innerHTML = '<%= price_tag(@invoice.cost) %>';
```
Comment (mayorova) [inline app/views/finance/provider/line_items/destroy.js.erb:11]: There is an issue that `invoice_cost_vat_amount` and the cell "Total Cost X Charged" which doesn't have an ID... but it is working like this in master, so - that's fine

File: app/javascript/src/Common/ajaxEvents.ts [hunk 14]
```diff
@@ -1,26 +1,66 @@
...
+
+/**
+ * Handle errors in formtastic forms rendered inside a colorbox. The error template is passed a
+ * responseText inside the XHR response.
+ *
+ * DEPRECATED: replace jquery/colorbox modals with Patternfly modals.
+ * TODO: add a colorbox specific selector to separate this with non-legacy implementations.
+ */
+$(document).on('ajax:error', 'form[data-remote]:not(.pf-c-form)', (event) => {
+  const [,, xhr] = (event.originalEvent as AJAXErrorEvent<HTMLDocument>).detail
```
Comment (mayorova) [inline app/javascript/src/Common/ajaxEvents.ts:59]:
```suggestion
  const [,, xhr] = (event.originalEvent as AJAXErrorEvent<Document>).detail
```

Just a TS suggestion...

```
/** @deprecated use Document */
interface HTMLDocument extends Document {
```

File: app/views/api/metric_visibilities/change.js.erb (hunk 6 above)
Comment (josemigallas) [reply app/views/api/metric_visibilities/change.js.erb:4]: The latter is a shorthand of the former, but they are the same.
Long form `render partial: ''` is unnecessary unless using a layout.
I tend to use it though.. because I like explicit, but I admit it's unnecessary verbose.

File: app/javascript/packs/provider.ts (hunk 9 above)
Comment (josemigallas) [reply app/javascript/packs/provider.ts:22]: Yes, `jquery_ujs` embedded it inside `window.$`, whereas `rails-ujs` declare in its own global scope.

File: app/javascript/src/Types/rails-ujs.d.ts (hunk 10 above)
Comment (josemigallas) [reply app/javascript/src/Types/rails-ujs.d.ts:4]: Yep 😂

File: app/views/finance/provider/line_items/destroy.js.erb (hunk 13 above)
Comment (josemigallas) [reply app/views/finance/provider/line_items/destroy.js.erb:11]: Yes, I created a JIRA to review this mess https://issues.redhat.com/browse/THREESCALE-11751

File: app/javascript/src/Common/ajaxEvents.ts [hunk 15]
```diff
@@ -18,47 +18,31 @@ import $ from 'jquery'
...
  *
  * DEPRECATED: replace jquery/colorbox modals with Patternfly modals.
  * TODO: add a colorbox specific selector to separate this with non-legacy implementations.
  */
 $(document).on('ajax:error', 'form[data-remote]:not(.pf-c-form)', (event) => {
-  const [,, xhr] = (event.originalEvent as AJAXErrorEvent<HTMLDocument>).detail
+  const [,, xhr] = (event.originalEvent as AJAXErrorEvent<Document>).detail
 
-  // If EventTarget is replaceWith'ed immediately, ajax:success and ajax:complete are not called and the spinner will stay
+  // If EventTarget (modal's content) is replaced immediately, ajax:success and ajax:complete are not called and the spinner will stay
```
Comment (mayorova) [inline app/javascript/src/Common/ajaxEvents.ts:45]: So, does `setTimeout()` wait for all the events to be called and handled?...

File: app/views/buyers/accounts/show.html.slim [hunk 16]
```diff
@@ -39,7 +39,7 @@ div class="pf-l-grid pf-m-gutter pf-m-all-6-col"
         div class="pf-l-flex__item"
           div class="pf-c-card"
             div class="pf-c-card__body"
-              = render partial: 'account_plan', locals: { plan: plan,contract:  @account.bought_account_contract }
+              = render partial: 'account_plan', locals: { plan:, contract:  @account.bought_account_contract }
```
Comment (mayorova) [inline app/views/buyers/accounts/show.html.slim:42]:
```suggestion
              = render partial: 'account_plan', locals: { plan:, contract: @account.bought_account_contract }
```

**Review by mayorova — DISMISSED** [review]

File: app/javascript/src/Common/ajaxEvents.ts (hunk 15 above)
Comment (josemigallas) [reply app/javascript/src/Common/ajaxEvents.ts:45]: setTimeout 0 waits for the next cycle, which will happen after the current pile of events handlers are executed (due to JS being single thread).

File: app/views/buyers/accounts/show.html.slim (hunk 16 above)
Comment (josemigallas) [reply app/views/buyers/accounts/show.html.slim:42]: 💩!!

File: app/views/api/features/create.js.erb (hunk 7 above)
Comment (josemigallas) [reply app/views/api/features/create.js.erb]: Fair enough, server is unreliable for this evident client-side responsibility. But I don't like to run this line every time needlessly, so I pass the ball to CSS.

Comment (mayorova) [reply app/views/api/features/create.js.erb]: I think it's a good solution!
